File: ZipFlash/app/src/main/java/com/zipflash/mrp/helper/UpdateChecker.java
package com.zipflash.mrp.helper;

import android.app.Activity;
import android.app.AlertDialog;
import android.app.DownloadManager;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.database.Cursor;
import android.net.Uri;
import android.os.Environment;
import android.os.Handler;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.TextView;

import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.File;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;
import com.zipflash.mrp.R;
import android.content.pm.PackageManager;
import android.content.ActivityNotFoundException;
import android.widget.Toast;

public class UpdateChecker {

    private Activity mActivity;
    private String mJsonUrl;
    private AlertDialog progressDialog; // For showing download progress
    private static final String APK_FILENAME = "ZipFlash-update.apk";

    public UpdateChecker(Activity activity, String jsonUrl) {
        mActivity = activity;
        mJsonUrl = jsonUrl;
    }

    public void checkForUpdate(final String currentVersion) {
        new Thread(new Runnable() {
				@Override
				public void run() {
					try {
						URL url = new URL(mJsonUrl);
						HttpURLConnection conn = (HttpURLConnection) url.openConnection();
						conn.setRequestMethod("GET");
						conn.setConnectTimeout(5000);
						conn.setReadTimeout(5000);
						conn.connect();

						InputStream is = conn.getInputStream();
						BufferedReader reader = new BufferedReader(new InputStreamReader(is));
						StringBuilder sb = new StringBuilder();
						String line;
						while ((line = reader.readLine()) != null) {
							sb.append(line);
						}
						reader.close();
						is.close();
						conn.disconnect();

						JSONObject json = new JSONObject(sb.toString());
						final String latestVersion = json.getString("latest_version");
						final String apkUrl = json.getString("apk_url");
						final String changelog = json.optString("changelog", "");

						// Compare versions
						if (!currentVersion.equals(latestVersion)) {
							mActivity.runOnUiThread(new Runnable() {
									@Override
									public void run() {
										showUpdateDialog(latestVersion, changelog, apkUrl);
									}
								});
						}

					} catch (final Exception e) {
						e.printStackTrace();
						Log.e("UpdateChecker", "Update check failed: " + e.getMessage());
					}
				}
			}).start();
    }

    private void showUpdateDialog(final String latestVersion, String changelog, final String apkUrl) {
        final AlertDialog dialog = new AlertDialog.Builder(mActivity, R.style.CustomDialogTheme)
			.create();

        dialog.setCancelable(false); // User must update

        // Inflate custom layout
        LinearLayout layout = (LinearLayout) mActivity.getLayoutInflater()
			.inflate(R.layout.dialog_update, null);

        TextView tvTitle = layout.findViewById(R.id.tvTitle);
        TextView tvChangelog = layout.findViewById(R.id.tvChangelog);
        Button btnUpdate = layout.findViewById(R.id.btnUpdateNow);

        tvTitle.setText("Update Available (" + latestVersion + ")");
        tvChangelog.setText(changelog);

        btnUpdate.setOnClickListener(new View.OnClickListener() {
				@Override
				public void onClick(View v) {
					dialog.dismiss();
					downloadApk(apkUrl);
				}
			});

        dialog.setView(layout);
        dialog.show();

        // Make background transparent
        if (dialog.getWindow() != null) {
            dialog.getWindow().setBackgroundDrawable(
				new android.graphics.drawable.ColorDrawable(android.graphics.Color.TRANSPARENT)
            );
        }
    }

    private void downloadApk(String url) {
        // Check and delete existing APK to ensure override
        File downloadsDir = Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS);
        File existingApk = new File(downloadsDir, APK_FILENAME);
        if (existingApk.exists()) {
            boolean deleted = existingApk.delete();
            if (!deleted) {
                Log.w("UpdateChecker", "Failed to delete existing APK: " + existingApk.getAbsolutePath());
            }
        }

        // Create a ProgressDialog with a ProgressBar
        progressDialog = new AlertDialog.Builder(mActivity, R.style.CustomDialogTheme)
			.create();
        progressDialog.setCancelable(false); // Prevent dismissing during download

        // Inflate custom layout for progress dialog
        LinearLayout layout = (LinearLayout) mActivity.getLayoutInflater()
			.inflate(R.layout.dialog_progress, null);

        final TextView tvProgress = layout.findViewById(R.id.tvProgress);
        final ProgressBar progressBar = layout.findViewById(R.id.progressBar);

        tvProgress.setText("Downloading update...");
        progressBar.setMax(100); // Set max progress to 100%

        progressDialog.setView(layout);
        progressDialog.show();

        // Make background transparent
        if (progressDialog.getWindow() != null) {
            progressDialog.getWindow().setBackgroundDrawable(
				new android.graphics.drawable.ColorDrawable(android.graphics.Color.TRANSPARENT)
            );
        }

        DownloadManager.Request request = new DownloadManager.Request(Uri.parse(url));
        request.setTitle("Downloading update...");
        request.setDescription("ZipFlash APK update");
        request.setNotificationVisibility(DownloadManager.Request.VISIBILITY_VISIBLE_NOTIFY_COMPLETED);
        request.setDestinationInExternalPublicDir(Environment.DIRECTORY_DOWNLOADS, APK_FILENAME);

        final DownloadManager manager = (DownloadManager) mActivity.getSystemService(Context.DOWNLOAD_SERVICE);
        final long downloadId = manager.enqueue(request);

        // Check download progress
        final Handler handler = new Handler();
        final Runnable progressRunnable = new Runnable() {
            @Override
            public void run() {
                DownloadManager.Query query = new DownloadManager.Query();
                query.setFilterById(downloadId);
                Cursor cursor = manager.query(query);
                if (cursor != null && cursor.moveToFirst()) {
                    int status = cursor.getInt(cursor.getColumnIndex(DownloadManager.COLUMN_STATUS));
                    if (status == DownloadManager.STATUS_SUCCESSFUL) {
                        final String uriString = cursor.getString(cursor.getColumnIndex(DownloadManager.COLUMN_LOCAL_URI));
                        cursor.close();
                        progressDialog.dismiss();
                        // Automatically start APK installation
                        mActivity.runOnUiThread(new Runnable() {
								@Override
								public void run() {
									installApk(uriString);
								}
							});
                    } else if (status == DownloadManager.STATUS_RUNNING) {
                        // Calculate progress
                        long bytesDownloaded = cursor.getLong(cursor.getColumnIndex(DownloadManager.COLUMN_BYTES_DOWNLOADED_SO_FAR));
                        long bytesTotal = cursor.getLong(cursor.getColumnIndex(DownloadManager.COLUMN_TOTAL_SIZE_BYTES));
                        if (bytesTotal > 0) {
                            final int progress = (int) ((bytesDownloaded * 100L) / bytesTotal);
                            mActivity.runOnUiThread(new Runnable() {
									@Override
									public void run() {
										progressBar.setProgress(progress);
										tvProgress.setText("Downloading: " + progress + "%");
									}
								});
                        }
                        handler.postDelayed(this, 500); // Update every 500ms
                    } else if (status == DownloadManager.STATUS_FAILED) {
                        cursor.close();
                        progressDialog.dismiss();
                        Log.e("UpdateChecker", "Download failed");
                        mActivity.runOnUiThread(new Runnable() {
								@Override
								public void run() {
									new AlertDialog.Builder(mActivity)
                                        .setTitle("Download Failed")
                                        .setMessage("Failed to download the update. Please try again.")
                                        .setPositiveButton("OK", new DialogInterface.OnClickListener() {
                                            @Override
                                            public void onClick(DialogInterface dialog, int which) {
                                                dialog.dismiss();
                                            }
                                        })
                                        .show();
								}
							});
                    } else {
                        handler.postDelayed(this, 500); // Retry in 500ms
                    }
                } else {
                    if (cursor != null) {
                        cursor.close();
                    }
                    handler.postDelayed(this, 500); // Retry in 500ms
                }
            }
        };
        handler.post(progressRunnable);
    }

    private void installApk(String uriString) {
        Uri apkUri = null;
        File apkFile = null;

        // Try to use the DownloadManager's URI first
        if (uriString != null) {
            try {
                apkUri = Uri.parse(uriString);
                apkFile = new File(apkUri.getPath());
                if (!apkFile.exists()) {
                    apkUri = null; // File doesn't exist, fallback to manual search
                }
            } catch (Exception e) {
                Log.e("UpdateChecker", "Invalid URI: " + e.getMessage());
                apkUri = null; // Fallback to manual search
            }
        }

        // Fallback: Check Downloads folder for the APK
        if (apkUri == null) {
            File downloadsDir = Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS);
            apkFile = new File(downloadsDir, APK_FILENAME);
            if (apkFile.exists()) {
                apkUri = Uri.fromFile(apkFile);
            }
        }

        // Attempt to install the APK
        if (apkUri != null && apkFile != null && apkFile.exists()) {
            try {
                Intent intent = new Intent(Intent.ACTION_VIEW);
                intent.setDataAndType(apkUri, "application/vnd.android.package-archive");
                intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_GRANT_READ_URI_PERMISSION);
                mActivity.startActivity(intent);
            } catch (Exception e) {
                Log.e("UpdateChecker", "Failed to start APK installation: " + e.getMessage());
                showInstallationError();
            }
        } else {
            Log.e("UpdateChecker", "APK file not found in Downloads folder");
            showInstallationError();
        }
    }

	private void showInstallationError() {
		mActivity.runOnUiThread(new Runnable() {
				@Override
				public void run() {
					AlertDialog dialog = new AlertDialog.Builder(mActivity, R.style.CustomDialogTheme)
						.setTitle("Download Complete!")
						.setMessage("Please install it manually from the Downloads folder or check your Notification panel.")
						.setPositiveButton("Close ZipFlash", new DialogInterface.OnClickListener() {
							@Override
							public void onClick(DialogInterface dialog, int which) {
								dialog.dismiss();
								// Try to open ZArchiver's Download folder
								Intent zArchiverIntent = new Intent(Intent.ACTION_VIEW);
								zArchiverIntent.setPackage("ru.zdevs.zarchiver");
								Uri uri = Uri.fromFile(Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS));
								zArchiverIntent.setDataAndType(uri, "resource/folder");
								zArchiverIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);

								// Check if ZArchiver is installed
								PackageManager pm = mActivity.getPackageManager();
								if (zArchiverIntent.resolveActivity(pm) != null) {
									try {
										mActivity.startActivity(zArchiverIntent);
									} catch (ActivityNotFoundException e) {
										openGenericFileManager();
									}
								} else {
									openGenericFileManager();
								}
							}
						})
						.create();

					dialog.setCancelable(false); // Prevent dismissal via back button
					dialog.setCanceledOnTouchOutside(false); // Prevent dismissal by touching outside
					dialog.show();
				}

				

				private void openGenericFileManager() {
					Intent intent = new Intent(Intent.ACTION_VIEW);
					Uri uri = Uri.fromFile(Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS));
					intent.setDataAndType(uri, "resource/folder");
					intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
					try {
						mActivity.startActivity(Intent.createChooser(intent, "Open Download Folder"));
					} catch (ActivityNotFoundException e) {
						Toast.makeText(mActivity.getApplicationContext(), "No file manager found to open Download folder", Toast.LENGTH_SHORT).show();
					}
				}
			});
	}
    
}

File: ZipFlash/app/src/main/java/com/zipflash/mrp/helper/UpdateChecker.kt
package com.zipflash.mrp.helper

import android.app.Activity
import android.app.AlertDialog
import android.app.DownloadManager
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Environment
import android.os.Handler
import android.os.Looper
import android.util.Log
import android.view.View
import android.widget.Button
import android.widget.LinearLayout
import android.widget.ProgressBar
import android.widget.TextView
import android.widget.Toast
import com.zipflash.mrp.R
import com.zipflash.mrp.api.ApiClient
import com.zipflash.mrp.helper.PermissionHelper
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import java.io.File
import java.lang.ref.WeakReference

class UpdateChecker(activity: Activity, private val mJsonUrl: String = GITHUB_UPDATE_URL) {

    companion object {
        const val GITHUB_UPDATE_URL = "https://raw.githubusercontent.com/marterp/ZipFlash-NoRoot/refs/heads/main/update.json"
        const val APK_FILENAME = "ZipFlash-update.apk"
    }

    private val mActivityRef = WeakReference(activity)
    private var progressDialog: AlertDialog? = null
    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())
    private var downloadJob: kotlinx.coroutines.Job? = null

    private fun getActivity(): Activity? = mActivityRef.get()

    fun checkForUpdate(currentVersion: String) {
        scope.launch {
            try {
                val call = ApiClient.updateApi.checkUpdateSync()
                val response = call.execute().body() ?: throw Exception("Empty response")
                val latestVersion = response.latestVersion
                val apkUrl = response.apkUrl
                val changelog = response.changelog

                if (currentVersion != latestVersion) {
                    val act = getActivity()
                    act?.runOnUiThread { showUpdateDialog(latestVersion, changelog, apkUrl) }
                }
            } catch (e: Exception) {
                e.printStackTrace()
                Log.e("UpdateChecker", "Update check failed: ${e.message}")
            }
        }
    }

    private fun showUpdateDialog(latestVersion: String, changelog: String, apkUrl: String) {
        val act = getActivity() ?: return

        val dialog = AlertDialog.Builder(act, R.style.CustomDialogTheme).create()
        dialog.setCancelable(false)

        val layout = act.layoutInflater.inflate(R.layout.dialog_update, null) as LinearLayout
        val tvTitle = layout.findViewById<TextView>(R.id.tvTitle)
        val tvChangelog = layout.findViewById<TextView>(R.id.tvChangelog)
        val btnUpdate = layout.findViewById<Button>(R.id.btnUpdateNow)

        tvTitle.text = "Update Available ($latestVersion)"
        tvChangelog.text = changelog

        btnUpdate.setOnClickListener {
            dialog.dismiss()
            downloadApk(apkUrl)
        }

        dialog.setView(layout)
        dialog.show()

        dialog.window?.setBackgroundDrawable(android.graphics.drawable.ColorDrawable(android.graphics.Color.TRANSPARENT))
    }

    private fun downloadApk(url: String) {
        val act = getActivity() ?: return

        if (!PermissionHelper.hasNotificationPermission(act)) {
            PermissionHelper.requestNotificationPermission(act)
        }

        val downloadsDir = Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS)
        val existingApk = File(downloadsDir, APK_FILENAME)
        if (existingApk.exists()) {
            val deleted = existingApk.delete()
            if (!deleted) {
                Log.w("UpdateChecker", "Failed to delete existing APK: ${existingApk.absolutePath}")
            }
        }

        progressDialog = AlertDialog.Builder(act, R.style.CustomDialogTheme).create()
        progressDialog?.setCancelable(false)

        val layout = act.layoutInflater.inflate(R.layout.dialog_progress, null) as LinearLayout
        val tvProgress = layout.findViewById<TextView>(R.id.tvProgress)
        val progressBar = layout.findViewById<ProgressBar>(R.id.progressBar)

        tvProgress.text = "Downloading update..."
        progressBar.max = 100

        progressDialog?.setView(layout)
        progressDialog?.show()

        progressDialog?.window?.setBackgroundDrawable(android.graphics.drawable.ColorDrawable(android.graphics.Color.TRANSPARENT))

        val request = DownloadManager.Request(Uri.parse(url))
        request.setTitle("Downloading update...")
        request.setDescription("ZipFlash APK update")
        request.setNotificationVisibility(DownloadManager.Request.VISIBILITY_VISIBLE_NOTIFY_COMPLETED)
        request.setDestinationInExternalPublicDir(Environment.DIRECTORY_DOWNLOADS, APK_FILENAME)

        val manager = act.getSystemService(Context.DOWNLOAD_SERVICE) as DownloadManager
        val downloadId = manager.enqueue(request)

        downloadJob = scope.launch {
            while (isActive) {
                val query = DownloadManager.Query()
                query.setFilterById(downloadId)
                val cursor = manager.query(query)
                if (cursor != null && cursor.moveToFirst()) {
                    val statusIndex = cursor.getColumnIndex(DownloadManager.COLUMN_STATUS)
                    if (statusIndex < 0) {
                        delay(500)
                        continue
                    }
                    val status = cursor.getInt(statusIndex)
                    if (status == DownloadManager.STATUS_SUCCESSFUL) {
                        val uriIndex = cursor.getColumnIndex(DownloadManager.COLUMN_LOCAL_URI)
                        val uriString = if (uriIndex >= 0) cursor.getString(uriIndex) else null
                        cursor.close()
                        progressDialog?.dismiss()
                        val act = getActivity()
                        act?.runOnUiThread { installApk(uriString) }
                        break
                    } else if (status == DownloadManager.STATUS_RUNNING) {
                        val bytesDownloadedIndex = cursor.getColumnIndex(DownloadManager.COLUMN_BYTES_DOWNLOADED_SO_FAR)
                        val bytesTotalIndex = cursor.getColumnIndex(DownloadManager.COLUMN_TOTAL_SIZE_BYTES)
                        if (bytesDownloadedIndex >= 0 && bytesTotalIndex >= 0) {
                            val bytesDownloaded = cursor.getLong(bytesDownloadedIndex)
                            val bytesTotal = cursor.getLong(bytesTotalIndex)
                            if (bytesTotal > 0) {
                                val progress = ((bytesDownloaded * 100L) / bytesTotal).toInt()
                                val act = getActivity()
                                act?.runOnUiThread {
                                    progressBar.progress = progress
                                    tvProgress.text = "Downloading: $progress%"
                                }
                            }
                        }
                        delay(500)
                    } else if (status == DownloadManager.STATUS_FAILED) {
                        cursor.close()
                        progressDialog?.dismiss()
                        Log.e("UpdateChecker", "Download failed")
                        val act = getActivity()
                        act?.runOnUiThread {
                            AlertDialog.Builder(act)
                                .setTitle("Download Failed")
                                .setMessage("Failed to download the update. Please try again.")
                                .setPositiveButton("OK") { dialog, _ -> dialog.dismiss() }
                                .show()
                        }
                        break
                    } else {
                        delay(500)
                    }
                } else {
                    delay(500)
                }
            }
        }
    }

    fun cancelDownload() {
        downloadJob?.cancel()
        downloadJob = null
        progressDialog?.dismiss()
    }

    private fun installApk(uriString: String?) {
        val act = getActivity() ?: return

        var apkUri: Uri? = null
        var apkFile: File? = null

        if (uriString != null) {
            try {
                apkUri = Uri.parse(uriString)
                apkFile = File(apkUri.path!!)
                if (!apkFile.exists()) {
                    apkUri = null
                }
            } catch (e: Exception) {
                Log.e("UpdateChecker", "Invalid URI: ${e.message}")
                apkUri = null
            }
        }

        if (apkUri == null) {
            val downloadsDir = Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS)
            apkFile = File(downloadsDir, APK_FILENAME)
            if (apkFile.exists()) {
                apkUri = Uri.fromFile(apkFile)
            }
        }

        if (apkUri != null && apkFile != null && apkFile.exists()) {
            try {
                val intent = Intent(Intent.ACTION_VIEW)
                intent.setDataAndType(apkUri, "application/vnd.android.package-archive")
                intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_GRANT_READ_URI_PERMISSION
                act.startActivity(intent)
            } catch (e: Exception) {
                Log.e("UpdateChecker", "Failed to start APK installation: ${e.message}")
                showInstallationError()
            }
        } else {
            Log.e("UpdateChecker", "APK file not found in Downloads folder")
            showInstallationError()
        }
    }

    private fun showInstallationError() {
        val act = getActivity() ?: return

        act.runOnUiThread {
            val dialog = AlertDialog.Builder(act, R.style.CustomDialogTheme)
                .setTitle("Download Complete!")
                .setMessage("Please install it manually from the Downloads folder or check your Notification panel.")
                .setPositiveButton("Close ZipFlash") { dialog, _ ->
                    dialog.dismiss()
                    val zArchiverIntent = Intent(Intent.ACTION_VIEW)
                    zArchiverIntent.setPackage("ru.zdevs.zarchiver")
                    val uri = Uri.fromFile(Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS))
                    zArchiverIntent.setDataAndType(uri, "resource/folder")
                    zArchiverIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)

                    val pm = act.packageManager
                    if (zArchiverIntent.resolveActivity(pm) != null) {
                        try {
                            act.startActivity(zArchiverIntent)
                        } catch (e: android.content.ActivityNotFoundException) {
                            openGenericFileManager()
                        }
                    } else {
                        openGenericFileManager()
                    }
                }
                .create()

            dialog.setCancelable(false)
            dialog.setCanceledOnTouchOutside(false)
            dialog.show()
        }
    }

    private fun openGenericFileManager() {
        val act = getActivity() ?: return
        val intent = Intent(Intent.ACTION_VIEW)
        val uri = Uri.fromFile(Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS))
        intent.setDataAndType(uri, "resource/folder")
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        try {
            act.startActivity(Intent.createChooser(intent, "Open Download Folder"))
        } catch (e: android.content.ActivityNotFoundException) {
            Toast.makeText(act.applicationContext, "No file manager found to open Download folder", Toast.LENGTH_SHORT).show()
        }
    }
}
